### PR TITLE
Rewrites PLUS and BITWISE_AND implementations using new modeling

### PR DESCRIPTION
**Pull request by johnedquinn**
## Relevant Issues
- #1646 

## Description
- Rewrites PLUS and BITWISE_AND
  - Slightly updates how function instances are chosen. See FnResolver. 
  - See `ArithmeticDiadicOperator`. It holds a table of pointers to functions given two argument types. The function is in charge of providing the closest function impl, but the engine is still in charge of coercions and final resolution.
- Passes all tests, and now is even more accurate. See the new `plusTests` in `PartiQLEvaluatorTests`
- Dynamic calls now hold functions, not functions instances since they haven't been resolved yet.

## Performance Implications
This _should_ make dynamic dispatches faster, since we now no longer need to loop over several candidates. They are consolidated. On top of that, the choosing of an implementation is internalized and uses a lookup table to find the most appropriate instance for PLUS and BITWISE_AND.

Edit: I ran some quick benchmarks to confirm this theory, and the assumption is correct. I've published these benchmarks via a [tag](https://github.com/johnedquinn/partiql-lang-jvm/releases/tag/v1-plus-optimization) on my fork. Comparing the new plus operator and the old minus operator:

```
Benchmark                               Mode  Cnt  Score    Error  Units
PartiQLBenchmark.minusIntDynamicStatic  avgt   20  0.150 ±  0.005  us/op
PartiQLBenchmark.minusIntStatic         avgt   20  0.010 ±  0.001  us/op
PartiQLBenchmark.plusIntDynamicStatic   avgt   20  0.081 ±  0.005  us/op
PartiQLBenchmark.plusIntStatic          avgt   20  0.010 ±  0.001  us/op
```

The static execution time remains the same (makes sense), but the dynamic invocation sees performance gains. An approximate speedup of 1.85. This translates to almost twice as fast. And, looking at the implementation of `ExprCallDynamic`, I can see how we can make this even faster 👍 .

Huge props to @RCHowell for this modelling change.

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/TypePrecedence.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package org.partiql.spi.function.builtins
+
+import org.partiql.types.PType.Kind
+
+/**
+ * @return the precedence of the types for the PartiQL comparator.
+ * @see .TYPE_PRECEDENCE
+ */
+@Suppress("deprecation")
+internal val TYPE_PRECEDENCE: Map<Kind, Int> = listOf(
```
Comment (RCHowell) [inline partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/TypePrecedence.kt:10]: Top-level interval vals are ok afaik (functions were the leaked apis), but wouldn't hurt to put in a TypePrecedence object as a static field, kotlin is doing that anyways so let's at least control it.

**Review by RCHowell — CHANGES_REQUESTED** [review]

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnCollAgg.kt [hunk 2]
```diff
@@ -32,6 +32,7 @@ internal abstract class Fn_COLL_AGG__BAG__ANY(
     override fun getInstance(args: Array<PType>): Function.Instance = instance
 
     private val instance = object : Function.Instance(
+        name,
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnCollAgg.kt:35]: We have some logic checking the contents of the signature. Specifically, checking for `is_struct`.

File: partiql-planner/src/main/kotlin/org/partiql/planner/internal/CoercionFamily.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
+package org.partiql.planner.internal
+
+import org.partiql.types.PType
+
+/**
+ * This represents SQL:1999 Section 4.1.2 "Type conversions and mixing of data types" and breaks down the different
+ * coercion groups.
+ *
+ * TODO: [UNKNOWN] should likely be removed in the future. However, it is needed due to literal nulls and missings.
+ * TODO: [DYNAMIC] should likely be removed in the future. This is currently only kept to map function signatures.
```
[the remark this answers is not in the document]
Comment (johnedquinn) [reply partiql-planner/src/main/kotlin/org/partiql/planner/internal/CoercionFamily.kt:10]: Yeah, I talked to Josh about this last week -- how our variant is more similar to dynamic than say Spark's variant (which requires explicit casts). I'll leave this for now, but I think I'm inclined to agree.

File: partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnCollAgg.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (RCHowell) [reply partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnCollAgg.kt:35]: The IS_<TYPE> is out of spec and would be its own node anyways. Other than is_struct, is there a reason why instances would need names? 

I can see it being useful for introspection. I don't think it should be removed, but I don't think it should be user-provided.

**Review by RCHowell — APPROVED** [review]

**Pull request merged**